llvm/lib/CodeGen/AsmPrinter/DwarfDebug.h

```diff
@@ -409,6 +408,11 @@ class DwarfDebug : public DebugHandlerBase {
       std::pair<std::unique_ptr<DwarfTypeUnit>, const DICompositeType *>, 1>
       TypeUnitsUnderConstruction;
 
+  /// Used to set a uniqe ID for a Type Unit.
+  /// This counter represents number of DwarfTypeUnits created, not necessarily
+  /// number of type units that will be emitted.
+  unsigned NumTypeUnitsCreated = 0;
```

Comment:
It is used to create a unique ID for each type unit that is created. Which is used to create a "link" between DWARF5AccelTableData and TU it belongs to and TypeUnitMetaInfo. In emitDWARF5AccelTable we iterate over TUSymbols and create a mapping between unique ID and an index in LocalTU.

For CUs it's easier because unique ID just an index in whatever list they are contained. For TUs we 1) can abort TU creation 2) after it was fully construct it disregard it after writing it out. 

Previously this was that static counter.

Reply:
For TUs we try to construct them, but might not succeed. At which point we re-do everything but construct them in the CU. All the instances of TypeUnits that were created are discarded. If we do succeed we will write out TUs in to .debug_info section (for DWARF5), and then release the memory for all the TypeUnit instances. Thus this counter represents all the instances of the TypeUnits that were created, but not necessarily all the TypeUnits that were written out. It is used to assign a unique ID to each instance of TypeUnits which is referenced by DWARF5AcceleratorTable Entry, and is also used in TUSymbols to map being symbol to this ID.

When we write out entries that are for types we use to find in map index in LocalTU.

Reply:
Oh, sorry, yes I get what you're referring to.

What I was wondering is if we could assign the numbering later, once the type is being committed (when we're copying the accelerator table entries over from the temporary table to the main table & switching from DIE pointers to offsets - at that point we know the type units are real, and we could assign numbering there instead? (like the order/index of the entry in the TUSymbols list))